src/draw/lv_img_decoder.c

```diff
@@ -49,169 +29,93 @@ static lv_res_t lv_img_decoder_built_in_line_indexed(lv_img_dec_dsc_t * dsc, lv_
-
-/**
- * Initialize the image decoder module
- */
+void lv_bin_init(void);
 void _lv_img_decoder_init(void)
 {
     _lv_ll_init(&LV_GC_ROOT(_lv_img_decoder_ll), sizeof(lv_img_dec_t));
 
-    lv_img_dec_t * decoder;
-
-    /*Create a decoder for the built in color format*/
-    decoder = lv_img_decoder_create();
-    LV_ASSERT_MALLOC(decoder);
-    if(decoder == NULL) {
-        LV_LOG_WARN("lv_img_decoder_init: out of memory");
-        return;
-    }
-
-    lv_img_decoder_set_info_cb(decoder, lv_img_decoder_built_in_info);
-    lv_img_decoder_set_open_cb(decoder, lv_img_decoder_built_in_open);
-    lv_img_decoder_set_read_line_cb(decoder, lv_img_decoder_built_in_read_line);
-    lv_img_decoder_set_close_cb(decoder, lv_img_decoder_built_in_close);
+    /*Must be the last one in the list, since it's not restrictive in what it accepts*/
+    lv_bin_init();
 }
 
-/**
- * Get information about an image.
- * Try the created image decoder one by one. Once one is able to get info that info will be used.
- * @param src the image source. E.g. file name or variable.
- * @param header the image info will be stored here
- * @return LV_RES_OK: success; LV_RES_INV: wasn't able to get info about the image
- */
-lv_res_t lv_img_decoder_get_info(const void * src, lv_img_header_t * header)
+void lv_img_dec_dsc_in_init(lv_img_dec_dsc_in_t * desc, const lv_img_src_t * src, lv_color_t * color,
+                            lv_point_t * size_hint)
 {
-    lv_memset_00(header, sizeof(lv_img_header_t));
+    lv_memset_00(desc, sizeof(*desc));
+    desc->src = src;
+    if(color) desc->color = *color;
+    if(size_hint) desc->size_hint = *size_hint;
+}
 
-    if(src == NULL) return LV_RES_INV;
 
-    lv_img_src_type_t src_type = lv_img_src_get_type(src);
-    if(src_type == LV_IMG_SRC_OBJ) {
-        lv_img_src_t * src_obj = ((lv_img_src_t *)src);
-        src_type = src_obj->type;
-        src = src_obj->data;
+lv_res_t lv_img_decoder_get_info(const lv_img_dec_dsc_in_t * dsc_in, lv_img_header_t * header)
+{
+    /*TODO: We should search the cache here, since we might already have a decoder with the information we need */
+    lv_img_dec_dsc_t dsc = {0};
+    lv_memcpy(&dsc.input, dsc_in, sizeof(*dsc_in));
+    if(lv_img_decoder_open(&dsc, LV_IMG_DEC_ONLYMETA) == LV_RES_OK) {
+        lv_memcpy(header, &dsc.header, sizeof(*header));
+        return LV_RES_OK;
     }
+    return LV_RES_INV;
+}
 
-    if(src_type == LV_IMG_SRC_VARIABLE) {
-        const lv_img_dsc_t * img_dsc = src;
-        if(img_dsc->data == NULL) return LV_RES_INV;
-    }
+lv_img_dec_t * lv_img_decoder_accept(const lv_img_src_t * src, uint8_t * caps)
+{
+    if(src == NULL) return NULL;
 
     lv_res_t res = LV_RES_INV;
     lv_img_dec_t * d;
     _LV_LL_READ(&LV_GC_ROOT(_lv_img_decoder_ll), d) {
-        if(d->info_cb) {
-            res = d->info_cb(d, src, header);
-            if(res == LV_RES_OK) break;
+        if(d->accept_cb) {
+            res = d->accept_cb(src, caps);
```

Comment:
If `caps` is NULL we pass a dummy caps to avoid error check and error in the decoders. 

```
uint8_t dummy_caps = 0;
if(caps == NULL) caps = &dummy_caps;
```